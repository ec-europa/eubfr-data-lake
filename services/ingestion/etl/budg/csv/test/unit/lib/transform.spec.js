import mapper from '../../../src/lib/transform';
import testRecord from '../../stubs/record.json';

describe(`DG BUDG CSV transformer`, () => {
  let result = {};

  beforeAll(() => {
    result = mapper(testRecord);
  });

  test(`Throws an error on empty input`, () => {
    expect(mapper).toThrow();
  });

  test(`Timeframe information is transformed into a single object field`, () => {
    expect(result.timeframe).toBeDefined();
  });

  test(`Timeframe's "from" field is present`, () => {
    expect(result.timeframe.from).toBeDefined();
  });

  test(`Timeframe's "to" field is present`, () => {
    expect(result.timeframe.to).toBeDefined();
  });

  test(`"Name" source field is transformed to "title"`, () => {
    expect(result.title).toBeDefined();
    expect(result.title).toBe(`Central European Green Corridors`);
  });

  test(`"Related links" source field is transformed to "related_links"`, () => {
    expect(result.related_links).toBeDefined();
    expect(result.related_links).toBeTruthy();
  });

  test(`"Related links" has URL values correctly split`, () => {
    expect(result.related_links[0].url).toBeDefined();
    expect(result.related_links[0].label).toBeDefined();
  });

  test(`"Project location latitude" and "Project location longitude" move to "project_locations"`, () => {
    expect(result.project_locations).toBeDefined();
    // "AT;HR;DE;SK;SI" => there should be 5 locations.
    expect(result.project_locations.length).toBe(5);
  });

  test(`Project location information is correctly mapped`, () => {
    const firstLocation = result.project_locations[0];
    expect(firstLocation).toMatchSnapshot();
  });
});

describe(`DG BUDG transformer handling incomplete or inconsistent data`, () => {
  const partialRecord = {
    Name: 'Bulgaria’s capital expands metro network',
    'Timeframe start': '1388530800',
    'Timeframe end': '1420066800',
    'Project location latitude': '42.73806663',
    'Project location longitude': '23.40014509',
    'Project country(ies)': '',
    'EC’s priorities': '',
    Coordinators: '',
    Partners: '',
    'Related links': '',
  };

  test(`Certain fields should always be present after transformation`, () => {
    const partialResult = mapper(partialRecord);
    expect(partialResult).toMatchSnapshot();
  });
});
