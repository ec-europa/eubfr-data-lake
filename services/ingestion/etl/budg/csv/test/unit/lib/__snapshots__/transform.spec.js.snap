// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DG BUDG CSV transformer Produces correct JSON output structure 1`] = `
Object {
  "coordinators": Array [
    Object {
      "address": null,
      "country": null,
      "name": "Eva Maria Plunger (VERBUND AG)",
      "region": null,
      "type": null,
      "website": null,
    },
  ],
  "cover_image": "https://ec.europa.eu/budget/euprojects/sites/euprojects/files/at_cegc_985x270_0.jpg",
  "description": "<p>Under the EU-funded Central European Green Corridors project, 115 fast-charging stations for electric cars have opened in Germany, Austria, Slovenia, Slovakia and Croatia – allowing cross-border electromobility with your e-car between Munich and Zagreb. Further network expansion, offers and services for customers and energy supply solutions also drive the project’s roll-out of electric transport. Partners include public sector bodies and private companies.</p>",
  "ec_priorities": Array [
    "Jobs, Growth and Investment",
  ],
  "eu_budget_contribution": 3562000,
  "partners": Array [
    Object {
      "address": null,
      "country": null,
      "name": "Bayern Innovativ",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " BMW",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " Nissan",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " Renault",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " Volkswagen",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " Republic of Slovenia",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " ZSE",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " Schrack Technik",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " GreenWay",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " City of Zagreb",
      "region": null,
      "type": null,
      "website": null,
    },
    Object {
      "address": null,
      "country": null,
      "name": " SMATRICS",
      "region": null,
      "type": null,
      "website": null,
    },
  ],
  "programme_name": "Connecting Europe Facility",
  "project_id": "8155",
  "project_locations": Array [
    Object {
      "location": Object {
        "lat": "48.211312000000",
        "lon": "16.367315000000",
      },
      "name": "AT",
    },
    Object {
      "location": Object {
        "lat": "45.803508000000",
        "lon": "16.017459000000",
      },
      "name": "HR",
    },
    Object {
      "location": Object {
        "lat": "49.449922000000",
        "lon": "11.103470000000",
      },
      "name": "DE",
    },
    Object {
      "location": Object {
        "lat": "48.143047000000",
        "lon": "17.124260000000",
      },
      "name": "SK",
    },
    Object {
      "location": Object {
        "lat": "46.048183000000",
        "lon": "14.501456000000",
      },
      "name": "SI",
    },
  ],
  "project_website": "http://www.cegc-project.eu/",
  "related_links": Array [
    Object {
      "label": "INEA",
      "url": "https://ec.europa.eu/inea/en/ten-t/ten-t-projects/projects-by-country/multi-country/2013-eu-92069-s",
    },
    Object {
      "label": "InvestEU",
      "url": "https://europa.eu/investeu/projects/central-european-green-corridors_en",
    },
  ],
  "results": Object {
    "available": null,
    "result": "Installation and operation of 115 fast-charging stations in 5 European countries,Development and implementation of electromobility and cross-border services for e-car users,Strategic planning for a European fast-charging network.Since the implementation of the Central European Green Corridors project, the European fast-charging network for electric vehicles has continued to grow. An additional 200 fast-charging stations – 180 in Italy and 20 in Austria – are developed. The charging networks in the various countries are compatible with each other. This gives even more citizens the chance to take advantage of cross-border electromobility.",
  },
  "timeframe": Object {
    "from": "2013-12-31T23:00:00.000Z",
    "to": "2014-12-31T23:00:00.000Z",
  },
  "title": "Central European Green Corridors",
}
`;

exports[`DG BUDG CSV transformer Project location information is correctly mapped 1`] = `
Object {
  "location": Object {
    "lat": "48.211312000000",
    "lon": "16.367315000000",
  },
  "name": "AT",
}
`;

exports[`DG BUDG transformer handling incomplete or inconsistent data Certain fields should always be present after transformation 1`] = `
Object {
  "coordinators": Array [
    Object {
      "address": null,
      "country": null,
      "name": "",
      "region": null,
      "type": null,
      "website": null,
    },
  ],
  "cover_image": undefined,
  "description": undefined,
  "ec_priorities": Array [
    "",
  ],
  "eu_budget_contribution": NaN,
  "partners": Array [
    Object {
      "address": null,
      "country": null,
      "name": "",
      "region": null,
      "type": null,
      "website": null,
    },
  ],
  "programme_name": undefined,
  "project_id": undefined,
  "project_locations": Array [
    Object {
      "location": Object {
        "lat": "42.73806663",
        "lon": "23.40014509",
      },
      "name": "",
    },
  ],
  "project_website": undefined,
  "related_links": Array [],
  "results": Object {
    "available": null,
    "result": undefined,
  },
  "timeframe": Object {
    "from": "2013-12-31T23:00:00.000Z",
    "to": "2014-12-31T23:00:00.000Z",
  },
  "title": "Bulgaria’s capital expands metro network",
}
`;
